`sam-add-tag-from-fastq`: adds a tag to each SAM record with the corresponding sequence from a FASTQ file.

`sam-count-secondary`: counts and adds a tag with the number of alignments for each read.

`sam-mark-dups`: finds SAM records with the same start and end position and marks them as duplicates.

`sam-tag-to-fasta`: extract a SAM tag for each record and print it in FASTA format (useful for tags with sequences).

`sam-tag-value-dist`: measures how many times each tag value is found in SAM file.

`sam-to-bed`: converts a SAM file to BED.

`sam-to-bedgraph`: converts a SAM file to BEDGRAPH.

`sam-to-bigwig`: converts a SAM file to BIGWIG.

`sam-to-collapsed`: finds SAM records with the same start and end position and keeps only one.

`sam-per-ref-count.py`: Count reads aligned on reference sequence in a SAM/BAM file.
